oresat_configs/scripts/pdo.py

```diff
@@ -0,0 +1,148 @@
+"""Tools for working with PDOs"""
+
+import time
+from argparse import ArgumentParser, Namespace
+from typing import Optional
+
+import canopen
+
+from .. import Consts, OreSatConfig
+
+PDO = "list or receive PDOs from the specified card"
+
+
+def build_parser(parser: ArgumentParser) -> ArgumentParser:
+    """Configures an ArgumentParser suitable for this script.
+
+    The given parser may be standalone or it may be used as a subcommand in another ArgumentParser.
+    """
+    parser.description = PDO
+    parser.add_argument(
+        "--oresat",
+        default=Consts.default().arg,
+        choices=[m.arg for m in Consts],
+        type=lambda x: x.lower().removeprefix("oresat"),
+        help="oresat mission, defaults to %(default)s",
+    )
+    parser.add_argument("card", help="card name")
+    parser.add_argument(
+        "--list", action="store_true", help="list PDOs expected for the particular card"
+    )
+    parser.add_argument(
+        "--bus", default="vcan0", help="CAN bus to listen on, defaults to %(default)s"
+    )
+    return parser
+
+
+def register_subparser(subparsers):
+    """Registers an ArgumentParser as a subcommand of another parser.
+
+    Intended to be called by __main__.py for each script. Given the output of add_subparsers(),
+    (which I think is a subparser group, but is technically unspecified) this function should
+    create its own ArgumentParser via add_parser(). It must also set_default() the func argument
+    to designate the entry point into this script.
+    See https://docs.python.org/3/library/argparse.html#sub-commands, especially the end of that
+    section, for more.
+    """
+    parser = build_parser(subparsers.add_parser("pdo", help=PDO))
+    parser.set_defaults(func=pdo_main)
+
+
+typenames = {
+    canopen.objectdictionary.BOOLEAN: "bool",
+    canopen.objectdictionary.INTEGER8: "i8",
+    canopen.objectdictionary.INTEGER16: "i16",
+    canopen.objectdictionary.INTEGER32: "i32",
+    canopen.objectdictionary.UNSIGNED8: "u8",
+    canopen.objectdictionary.UNSIGNED16: "u16",
+    canopen.objectdictionary.UNSIGNED32: "u32",
+    canopen.objectdictionary.REAL32: "f32",
+    canopen.objectdictionary.VISIBLE_STRING: "str",
+    canopen.objectdictionary.OCTET_STRING: "bytes",
+    canopen.objectdictionary.UNICODE_STRING: "ustr",
+    canopen.objectdictionary.DOMAIN: "domain",
+    # canopen.objectdictionary.INTEGER24: 'i24',
+    canopen.objectdictionary.REAL64: "f64",
+    canopen.objectdictionary.INTEGER64: "i64",
+    # canopen.objectdictionary.UNSIGNED24: 'u24',
```

Comment:
We do not use unicode str, int24, or uin24, remove them.

Reply:
This dict is for display names of standard CANopen types (at least as implemented through the `canopen` python package). This does not use them in the configs, and the configs can continue not using those types. However, not including those types here would increase the possible valid PDO mappings that this script could not not display correctly. It would reduce compliance with the CANopen standard.